File: YYReader/Parsing/NovelParserRegistry.swift
import Foundation

actor NovelParserRegistry {
    private let adapters: [any NovelSourceAdapter] = [
        QidiySourceAdapter(),
        GenericNovelAdapter()
    ]

    func parseChapterPage(_ document: LoadedHTML) throws -> ParsedChapterPage {
        guard let adapter = adapters.first(where: { $0.canHandle(document) }) else {
            throw NovelParsingError.noReadableContent
        }
        return try adapter.parseChapterPage(document)
    }

    func parseCatalogPage(_ document: LoadedHTML) throws -> ParsedBookCatalog {
        guard let adapter = adapters.first(where: { $0.canHandle(document) }) else {
            throw NovelParsingError.missingCatalog
        }
        return try adapter.parseCatalogPage(document)
    }
}

File: windows/YYReader.Windows.Core/Parsing/NovelParserRegistry.cs
namespace YYReader.Windows.Core.Parsing;

public sealed class NovelParserRegistry
{
    private readonly IReadOnlyList<INovelSourceAdapter> _adapters =
    [
        new QidiySourceAdapter(),
        new GenericNovelAdapter()
    ];

    public ParsedChapterPage ParseChapterPage(LoadedHtml document)
    {
        var adapter = _adapters.FirstOrDefault(candidate => candidate.CanHandle(document));
        return adapter?.ParseChapterPage(document)
            ?? throw new NovelParsingException(NovelParsingErrorKind.NoReadableContent);
    }

    public ParsedBookCatalog ParseCatalogPage(LoadedHtml document)
    {
        var adapter = _adapters.FirstOrDefault(candidate => candidate.CanHandle(document));
        return adapter?.ParseCatalogPage(document)
            ?? throw new NovelParsingException(NovelParsingErrorKind.MissingCatalog);
    }
}
